AI Product Description: Highlight Tooltips Functionality

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #9290
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Finally, now that the tasks are complete, this PR also releases the whole AI product description feature and updates the release note with it.

This PR shows or hides the tooltip based on these conditions, where all must be true:

1. Product description is empty
2. The tooltip is presented less than 3 times. (To avoid annoying the user)
3. Tooltip was not dismissed manually by tapping on the “Got it” button.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

#### a. test to ensure tooltip doesn't show up on product with existing description
1. Run app,
2. Go to products tab,
3. Open a product with existing description, ensure the highlights tooltip does not show up.

#### b. test to ensure tooltip only appears 3 times if not dismissed
1. Go back to products tab,
2. Open a product with no description, and ensure the highlights tooltip shows up,
3. Tap back, then go back into the same product above. Repeatedly do this and make sure the tooltip only shows up for two more times.

#### c. test to ensure tooltip does not appear once dismissed
1. Log out to reset preferences.
2. Log back in, go to products tab,
3. Open a product with no description, and ensure the highlights tooltip shows up,
4. Dismiss it using the "Got it" button,
5. Leave the product details, then go back to the same product. Make sure the highlight tooltip does not show up again.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

<img src="https://github.com/woocommerce/woocommerce-android/assets/266376/df4a0725-bf59-4430-addf-81ff78d83a6d" width="42%" />

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hafizrahman committed:
UI: Push the tooltip a bit higher up to cover the "learn more" text and match iOS design.

hafizrahman committed:
Refactor: The code went to a different direction re: tooltip so delete unneeded code.

hafizrahman committed:
Refactor: Update existing Tooltip class to match usage.

hafizrahman committed:
Refactor: Simplify buttonView.show() into just one function.

This involves setting `link` and `tooltip` parameters as nullable, and only showing them if not null.

Additionally this commit also allows for passing tooltip data (title, message, etc) dynamically, instead of hardcoding them.

hafizrahman committed:
Feat: Build and show tooltip if product has no description.

hafizrahman committed:
UI: Show placeholder texts for the tooltip layout XML.

hafizrahman committed:
Feat: Release the product description AI from behind its feature flag.

hafizrahman committed:
Update: Release notes.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/models/ProductProperty.kt:
```diff
@@ -124,11 +124,14 @@ sealed class ProductProperty(val type: Type) {
         val onClick: (() -> Unit)
     ) : ProductProperty(BUTTON) {
+        /**
+         * @param onPrimaryButtonClick Set as null for tooltip dismiss action.
```
Not entirely sure this is the best way to do this, but I'm not sure how else someone can pass a function that can dismiss the `PopupWindow` as it needs to be specifically be its instance's `dismiss()` function.

JorgeMucientes replied:
Since the tooltip is not really a general purpose UI component, I wouldn't worry too much about this for now. I actually think you can get rid of `val onPrimaryButtonClick: (() -> Unit)?` and simply dismiss the tooltip as the primary action. 
If in the future we need to customize this, I don't see an scenario where clicking on any of the tooltip CTAs doesn't involve dismissing the tooltip among other things.

hafizrahman replied:
That makes sense. Handled this on https://github.com/woocommerce/woocommerce-android/pull/9377/commits/53e70671c6e9c1e39350844aeabe9094f2c99194

hafizrahman committed:
Test: Update e2e to check for either tooltip or product detail's root before proceeding.

This is because one of them will show up depending on whether the product has description or not.

hafizrahman committed:
Test: Fix e2e test as the previous commit didn't fully work.

Also renames the tooltip's button id.

hafizrahman commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/e2e/screens/products/ProductListScreen.kt:
```diff
@@ -27,9 +27,12 @@ class ProductListScreen : Screen {
-        waitForElementToBeDisplayed(R.id.highlights_tooltip_root)
-        clickOn(R.id.tooltip_dismiss_button)
-        waitForElementToBeDisplayed(R.id.productDetail_root)
+        val displayedElement = waitForAnyElementToBeDisplayed(R.id.highlights_tooltip_root, R.id.productDetail_root)
+        displayedElement?.let {
+            if (it == R.id.highlights_tooltip_root) {
+                clickOn(R.id.tooltip_primary_button)
+            }
+        }
```
@woocommerce/mobile-ui-testing-squad 

A new behavior added in this PR is that a product list screen will display a tooltip popup if the product description is empty.

I found that if the popup shows up, then `waitForElementToBeDisplayed(R.id.productDetail_root)` never returns true, likely because the popup uses `PopupWindow` that's displayed above the product list view. So I'm adding this where it checks if the popup shows up, and if yes it will click the dismiss button first.

hafizrahman replied:
I ended up modifying this since thing seems to be flaky (works for me locally, but broke here). I ended up following Ondrej's solution by simply hiding the tooltip for testing.

tiagomar replied:
> I ended up modifying this since thing seems to be flaky (works for me locally, but broke here)

Yeah, timing issues tend to kick in when running in CI with limited resources. Nice move hiding it for testing!  👍

hafizrahman replied:
@tiagomar Thank you for checking!

JorgeMucientes commented:
Hey @hafizrahman I haven't finished reviewing yet but I noticed a behavior I believe we should fix. After dismissing the tooltip inside product detail, if I navigate back and forth between the different product sections the Tooltip keeps "coming back":

https://github.com/woocommerce/woocommerce-android/assets/2663464/4cbc2e2f-1ceb-4eb6-98ba-519360e3127e

I believe once the user dismissed the tooltip the first time we shouldn't show the tooltip again for that same product screen.

hafizrahman committed:
Feat: Add AppPrefs to save whether AI tooltip is already dismissed or not.

hafizrahman committed:
Feat: Add dismiss functionality that saves to AppPrefs.

hafizrahman committed:
Feat: Add check for whether tooltip is dismissed already.

hafizrahman committed:
Test: e2e test continues to be inconsistent, so to simplify, just hide tooltip when testing.

hafizrahman committed:
Feat: Extra conditional to only show tooltip 3 times if user doesn't dismiss it.

hafizrahman commented:
@JorgeMucientes thank you! I updated the showing criteria based on @selanthiraiyan 's information. I updated the testing instructions as now there are more cases to check.

hafizrahman committed:
Refactor: Just use `dismiss button` term instead of `primary button`.

hafizrahman committed:
Fix: A workaround to avoid the tooltip to be displayed twice on each load.

hafizrahman committed:
Fix: Should allow the tooltip to show up 3 times, not 2.

hafizrahman commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/e2e/helpers/util/Screen.kt:
```diff
@@ -62,6 +62,9 @@ open class Screen {
         AppPrefs.wasAIProductDescriptionPromoDialogShown = true
+
+        // also hide AI description tooltip to make test more simple
+        AppPrefs.isAIProductDescriptionTooltipDismissed = true
```
Following the same solution as the line above it, to simply hide the tooltip to not break e2e test.

hafizrahman committed:
Fix: This function does not need to return value.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/WCProductPropertyButtonView.kt:
```diff
@@ -2,67 +2,76 @@ package com.woocommerce.android.ui.products
     defStyle: Int = 0
 ) : ConstraintLayout(context, attrs, defStyle) {
+    companion object {
+        const val TOOLTIP_VERTICAL_OFFSET = -8
```
Just to pull the tooltip up a little bit, to cover up the Learn more text and make the look match iOS.

JorgeMucientes commented:
Excellent job @hafizrahman 🥇 The feature works as expected and the applied changes look good! 

Once the minor `detekt` indentation issue is fixed the PR is ready to be merged!

hafizrahman committed:
Fix: Detekt.

hafizrahman committed:
Fix: Unit test.
